File: tessellation/static/js/composition.js
import { log } from './log.js';

export const SIDES = [0, 1, 2, 3];
export const CONNECTIONS = [1, 2, 3, 4, 5];
const [LEFT, TOP, RIGHT, BOTTOM] = SIDES;

export function isImageComplete(image) {
  return (
    image.url !== null &&
    image.connections.every(connection => connection !== null)
  );
}

export function generateTiles(images) {
  const out = [];
  let i = 0;
  images.forEach(image => {
    if (isImageComplete(image)) {
      for (let rotation = 0; rotation < SIDES.length; rotation++) {
        out.push({
          image,
          rotation
        });
      }
    }
  });
  return out;
}

function findLeftTile(composition, col, row) {
  if (col === 0) {
    return null;
  }
  return composition[row][col - 1];
}

function findTopTile(composition, col, row) {
  if (row === 0) {
    return null;
  }
  return composition[row - 1][col];
}

function calcPrevCoords(row, col, width) {
  if (col === 0) {
    if (row === 0) {
      return [null, null];
    }
    return [row - 1, width - 1];
  }
  return [row, col - 1];
}

function calcNextCoords(row, col, width, height) {
  if (col === width - 1) {
    if (row === height - 1) {
      return [null, null];
    }
    return [row + 1, 0];
  }
  return [row, col + 1];
}

function getConnection(tile, i) {
  return tile.image.connections[(i + tile.rotation) % SIDES.length];
}

function findRequirements(composition, col, row) {
  const leftTile = findLeftTile(composition, col, row);
  const topTile = findTopTile(composition, col, row);
  const requirements = [
    [LEFT, leftTile ? getConnection(leftTile, RIGHT) : null],
    [TOP, topTile ? getConnection(topTile, BOTTOM) : null]
  ];
  return requirements;
}

function fits(tile, requirements) {
  let i, side, requirement;
  for (i = 0; i < requirements.length; i++) {
    [side, requirement] = requirements[i];
    if (requirement !== null && getConnection(tile, side) !== requirement) {
      return false;
    }
  }
  return true;
}

function chooseTile(stack, requirements, excl = []) {
  let i, tile;
  for (i = 0; i < stack.length; i++) {
    tile = stack[i];
    if (excl.indexOf(tile) === -1 && fits(tile, requirements)) {
      stack.splice(i, 1);
      stack.push(tile);
      return tile;
    }
  }
  return null;
}

export function generateComposition(tiles, [width, height]) {
  const composition = Array.from(Array(height), () => new Array(width));
  const tried = Array.from(Array(height), () =>
    Array.from(Array(width), () => [])
  );
  const stack = Array.from(tiles);
  let row = 0,
    col = 0,
    tile,
    i = 0;
  for (i = 0; i < 1000; i++) {
    log(`Choosing tile [${row}, ${col}]`);
    tile = chooseTile(
      stack,
      findRequirements(composition, col, row),
      tried[row][col]
    );
    tried[row][col].push(tile);
    if (tile !== null) {
      composition[row][col] = tile;
      [row, col] = calcNextCoords(row, col, width, height);
      if (row === null) {
        return composition;
      }
    } else {
      log('No fitting tile found, going one step back');
      [row, col] = calcPrevCoords(row, col, width, height);
      if (row === null) {
        return null;
      }
    }
  }
}

File: tessellation/composition.py
import itertools
import logging
import random
import time
from dataclasses import dataclass
from typing import Callable, Dict, Iterator, List, Optional, Sequence

from tessellation.models import BOTTOM, LEFT, RIGHT, SIDES, TOP, Image, Tile

logger = logging.getLogger(__name__)

TComposition = List[List[Optional[Tile]]]
TTried = List[List[List[Tile]]]
TStack = List[Tile]
TUpdateStackFunc = Callable[[TStack, int], None]


@dataclass
class Requirement:
    side: int
    connection: int


@dataclass
class UpdateStackFunc:
    label: str
    func: TUpdateStackFunc


UPDATE_STACK_FUNC_DEFAULT_NAME: str = 'push_current'

UPDATE_STACK_FUNCS: Dict[str, UpdateStackFunc] = {
    UPDATE_STACK_FUNC_DEFAULT_NAME: UpdateStackFunc(
        label='Degrade last used tile',
        func=lambda stack, i: stack.append(stack.pop(i)),
    ),
    'cycle_stack': UpdateStackFunc(
        label='Repeat sequence',
        func=lambda stack, i: stack.append(stack.pop(0)),
    ),
    'noop': UpdateStackFunc(
        label='Prefer one tile', func=lambda stack, i: None
    ),
    'shuffle_stack': UpdateStackFunc(
        label='Random', func=lambda stack, i: random.shuffle(stack)
    ),
    'unshift_current': UpdateStackFunc(
        label='Prefer last used tile',
        func=lambda stack, i: stack.insert(0, stack.pop(i)),
    ),
}


def reverse_digits(n: int) -> int:
    if n < 10:
        return n
    rev = 0
    while n > 0:
        rev = rev * 10 + (n % 10)
        n = n // 10
    return rev


def generate_tiles(
    images: Sequence[Image], allow_rotation: bool = True
) -> List[Tile]:
    if not allow_rotation:
        return [Tile(image=image, rotation=0) for image in images]
    return list(
        itertools.chain.from_iterable(
            (Tile(image=image, rotation=rotation) for rotation in SIDES)
            for image in images
        )
    )


def find_requirements(
    composition: TComposition, col: int, row: int
) -> Iterator[Requirement]:
    if col != 0:
        left_tile = composition[row][col - 1]
        if left_tile:
            yield Requirement(
                side=LEFT,
                connection=reverse_digits(left_tile.get_connection(RIGHT)),
            )
    if row != 0:
        top_tile = composition[row - 1][col]
        if top_tile:
            yield Requirement(
                side=TOP,
                connection=reverse_digits(top_tile.get_connection(BOTTOM)),
            )


def fits(tile: Tile, requirements: Sequence[Requirement]) -> bool:
    for requirement in requirements:
        if tile.get_connection(requirement.side) != requirement.connection:
            return False
    return True


def choose_tile(
    stack: TStack,
    requirements: Sequence[Requirement],
    excl: Sequence[Tile],
    update_stack_func: TUpdateStackFunc,
) -> Optional[Tile]:
    for i in range(len(stack)):
        tile = stack[i]
        if tile not in excl and fits(tile, requirements):
            update_stack_func(stack, i)
            return tile
    return None


def generate_composition(
    tiles: List[Tile],
    width: int,
    height: int,
    max_steps: int = pow(2, 18),
    update_stack_func_name: str = UPDATE_STACK_FUNC_DEFAULT_NAME,
) -> TComposition:
    logger.info(f'Generating composition {update_stack_func_name=}')
    t0 = time.time()
    composition: TComposition = [
        [None for _ in range(width)] for _ in range(height)
    ]
    if not tiles:
        return composition
    tried: TTried = [[[] for _ in range(width)] for _ in range(height)]
    stack = tiles.copy()
    update_stack_func: TUpdateStackFunc = UPDATE_STACK_FUNCS[
        update_stack_func_name
    ].func  # type: ignore
    i = 0
    row = 0
    col = 0
    while row < height and col < width:
        tile = choose_tile(
            stack,
            list(find_requirements(composition, col, row)),
            tried[row][col],
            update_stack_func,
        )
        if tile is not None:
            tried[row][col].append(tile)
            composition[row][col] = tile
            if col == width - 1:
                row += 1
                col = 0
            else:
                col += 1
        else:
            # logger.info(f'[{row}, {col}] No fitting tile found, going one step back')
            tried[row][col].clear()
            if col == 0:
                if row == 0:
                    raise Exception("Input tiles don't connect.")
                row -= 1
                col = width - 1
            else:
                col -= 1
        if i >= max_steps:
            raise Exception(
                f'Failed to calculate composition in {max_steps} steps.'
            )
        i += 1
    t1 = time.time()
    logger.info(f'Generated composition in {t1 - t0:.5f}ms')
    return composition
